py-allennlp: Submission of py-allennlp and dependencies

essandess opened the pull request:
py-allennlp: Submission of py-allennlp and dependencies

* devel/jsonnet
* python/py-allennlp
* python/py-conllu
* python/py-editdistance
* python/py-flask-cors
* python/py-ftfy
* python/py-jsonnet
* python/py-jsonpickle
* python/py-overrides
* python/py-tensorboardX
* python/py-w2n

#### Description

<!-- Note: it is best to make pull requests from a branch rather than from master -->

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [x] submission
- [ ] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on
<!-- Generate version information with this command in shell:
    echo "macOS $(sw_vers -productVersion) $(sw_vers -buildVersion)"; echo "Xcode $(xcodebuild -version | awk '{print $NF}' | tr '\n' ' ')"
-->
macOS 10.x
Xcode 8.x

###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL?
<!-- Please don't open a new Trac ticket if you are submitting a pull request. -->
- [x] checked your Portfile with `port lint`?
- [x] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port install`?
- [x] tested basic functionality of all binary files?

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

Note: `sudo port -vst install` queries for GitHub credentials during the install of some dependency:
```
Cloning into '/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_python_py-pytorch/py37-pytorch/work/pytorch-1.3.1/android/libs/fbjni'...
Username for 'https://github.com': 
```

essandess commented:
I believe that this is building successfully, but timing out on all build bots. See https://paste.z0k.xyz/028582d4e966

mf2k commented on devel/jsonnet/Portfile:
```diff
@@ -0,0 +1,39 @@
+
+compiler.cxx_standard 2011
+
+use_configure       no
```
You need to make sure you are [using the right compiler](https://trac.macports.org/wiki/UsingTheRightCompiler) and test or disable the `universal` variant.

essandess replied:
Yes, this is necessary, otherwise the build fails. Discussed on the dev mailing list.

mf2k replied:
It's not about the build failing or not. This means that Macports selected compiler usually doe snot get set properly and you need to make sure it does. Also, you have not addressed the `universal` variant. I don't see how it would work in that case.

mf2k replied:
Can you post a link to the discussion?

essandess replied:
https://lists.macports.org/pipermail/macports-dev/2019-November/041518.html

essandess replied:
I don’t know too much about MacPorts compilers or the `universal` variant so if you just tell me the right settings I’ll toss them in. I do believe based on the 🤗 transformers port dependencies that Google likes cxx11 so it wasn’t a surprise that jsonnet needs that here.

mf2k replied:
I don't know the right settings to fix the universal variant because it is dependent on this port build system. You have to build it and look at the errors and fix them. Otherwise you should at least turn off the non-working variant. 

`universal_variant no
`

mf2k replied:
Looking briefly at the thread, it is way more time and knowledge than I have to determine if what you are doing is the right solution. It appears that the thread ends without resolution.

essandess replied:
> It appears that the thread ends without resolution.

The build successfully compiles a working binary and libraries, so afaik it’s resolved.

essandess replied:
Note: this port select causes the build to fail:
```
port select --show clang
The currently selected version for 'clang' is 'mp-clang-9.0'. 
port select --show gcc
The currently selected version for 'clang' is 'mp-gcc-9. 
```

Setting clang and gcc to `none` works.

What’s the correct blacklist invocation?

mf2k replied:
The fact that `port select` is affecting the build proves that it is not [using the right compiler](https://trac.macports.org/wiki/UsingTheRightCompiler).

essandess replied:
The latest versions build without the compiler directive.

mf2k commented on python/py-jsonnet/Portfile:
```diff
@@ -0,0 +1,45 @@
+
+    pre-build {
+        system -W ${worksrcpath} "make"
+    }
```
I'm not understanding why you are doing this? This is the default build phase command.

mf2k replied:
Is this because the python portgroup changes the build phase? If so, just set the `build.cmd` manually instead.

`build.cmd make
`

essandess replied:
That will break the Python build.

For some reason, setup.py wants to see all the build binaries from make here.

I don’t know encounters about setup.py to explain why. But I know that make then running setup.py as a standard MacPorts python port works.

essandess replied:
See https://github.com/macports/macports-ports/blob/46fc86b7595cd4ca581efa7dce6ac3962f005ac6/_resources/port1.0/group/python-1.0.tcl#L272 for why you can't edit `build.cmd`.

The idea is to run `make`, then let the python build take over.

reneeotten commented on python/py-allennlp/Portfile:
```diff
@@ -0,0 +1,123 @@
+PortGroup           github 1.0
+PortGroup           python 1.0
+
+github.setup        allenai allennlp a8c44ae
```
don't specify a commit-hash, use a proper version instead

essandess replied:
The latest official version is behind on its dependencies, so 0.9.1-prerelease is the best option right now. 

I’m not 100% convinced that it’s ready either, but am mostly convinced based on my own usage. It’ll take a while to eat through these edits and get a working build, so perhaps official 0.9.1 will re ready by then.

Until then it has to be the latest github hash.

reneeotten replied:
I’m pretty sure the latest stable upstream release does work - correct? So why not package that for now, you should not package development versions unless in a -devel port.

 I am not sure I understand - what does your statement “is behind on its dependencies” mean?

essandess replied:
See https://github.com/allenai/allennlp/issues/3481

Latest official doesn’t work with transformers, other dependencies. It can be modded to work, though.

reneeotten commented on python/py-allennlp/Portfile:
```diff
@@ -0,0 +1,123 @@
+                    learning models on a wide variety of linguistic\
+                    tasks.
+
+homepage            http://www.allennlp.org/
```
https

essandess replied:
Thanks done. 🦃🦃🦃

reneeotten commented on python/py-allennlp/Portfile:
```diff
@@ -0,0 +1,123 @@
+
+if {${name} ne ${subport}} {
+    depends_build-append \
+                    port:py${python.version}-setuptools
```
incorrect dependency type as it installs ```console_scripts```

essandess replied:
??? I no grok.

Please post the corrected code.

reneeotten replied:
I have explained this in some of your other PRs before, the idea is to learn from that for future submissions not to simply copy-paste review suggestion. I am sure you should be able to understand this by now and figure out what change is needed...

essandess replied:
Sorry, I’m not still sure what you’re referring to. Repetition is the mother of learning, and copy-and-paste is also efficient so some suggested code would be appreciated, especially in this huge complicated PR with zillions of little interdependent bits.

essandess replied:
I assume you want `depends_lib` here, so adding that.

reneeotten commented on python/py-allennlp/Portfile:
```diff
@@ -0,0 +1,123 @@
+                    port:py${python.version}-pytest \
+                    port:py${python.version}-responses
+
+    # Fix allennlp's dependecies on deprecated names pytorch-pretrained-bert|pytorch-transformers
```
based on the discussion on this topic with upstream I don't feel confident that this is "production-ready"

essandess replied:
See above.

reneeotten commented on python/py-allennlp/Portfile:
```diff
@@ -0,0 +1,123 @@
+    test.cmd        ${prefix}/bin/python${python.branch}
+    test.pre_args   {-c 'from allennlp.commands.test_install import TestInstall; TestInstall()'}
+    test.target
+
```
shouldn't you just be able to run this with ```py.test-${python.branch}``` instead, that would be preferred I think.

essandess replied:
Yes, but `botocore` is behind on its dependencies too. See https://github.com/boto/botocore/issues/1872

This hackiness hacks around this issue, which causes a dependency failure.

We can unhack this when `botocore` gets around to fixing its dependency issues.

reneeotten replied:
I cannot quickly see how these two are related... But as long as this way testing actually works it’s fine with me.

reneeotten commented on python/py-allennlp/Portfile:
```diff
@@ -0,0 +1,123 @@
+
+    test.run        yes
+    test.env        PYTHONPATH=${worksrcpath}/build/lib
+    test.cmd        ${prefix}/bin/python${python.branch}
```
if this needs to be like this, the change it to ```${python.bin}```

essandess replied:
Thanks done. 🦃🦃🦃

reneeotten commented on python/py-catalogue/Portfile:
```diff
@@ -38,7 +38,12 @@ if {${name} ne ${subport}} {
                     port:py${python.version}-setuptools
 
-    depends_run-append \
+    if {${python.version} < "38"} {
```
no need for ```"``` around the number, no other Portfile does this

essandess replied:
Thanks done. 🦃🦃🦃

reneeotten commented on python/py-catalogue/Portfile:
```diff
@@ -38,7 +38,12 @@ if {${name} ne ${subport}} {
+    if {${python.version} < "38"} {
+        depends_build-append \
+                    port:py${python.version}-importlib-metadata
+    }
```
this is imported in the files installed by ```py-catalogue``` (i.e., ```/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/catalogue.py```). So this is for sure a runtime dependency. If it's needed at build-time as well you should add it as ```lib``, otherwise as ```run```.

essandess replied:
Thanks done. 🦃🦃🦃

reneeotten commented on python/py-catalogue/Portfile:
```diff
@@ -38,7 +38,12 @@ if {${name} ne ${subport}} {
+    }
+
+    depends_test-append \
                     port:py${python.version}-pytest
```
you're adding a test dependency, but no test-phase - that does not make sense. Please add a ```test``` phase so that the tests can be run.

essandess replied:
There’s some test script, but I don’t know how to invoke it from MacPorts. See https://github.com/explosion/catalogue/blob/master/test_catalogue.py

Please post code if you know the approach.

reneeotten replied:
what have you tried? By now you’ve encountered several varieties (i.e., using py.test-${python.branch} or ${python.bin}) - I am sure that one of those approaches will work.

essandess replied:
I haven’t enabled test.run because afaik there’s no magic that let’s it find the test script linked above. Is there? If so, I’ll just enable it and hope it works.

essandess replied:
Done thanks.

reneeotten commented on python/py-conllu/Portfile:
```diff
@@ -0,0 +1,48 @@
+PortGroup           github 1.0
+PortGroup           python 1.0
+
+github.setup        EmilStenstrom conllu 68199a7
```
don't specify a commit hash, get a version... It appears that upstream does not tag versions properly on GitHub. In that case get them from PyPI instead.

essandess replied:
Thanks done. 🦃🦃🦃

reneeotten commented on python/py-editdistance/Portfile:
```diff
@@ -0,0 +1,48 @@
+    test.env        PYTHONPATH=${worksrcpath}/build/lib
+    test.cmd        python${python.branch}
+    test.target     test/test_${github.project}.py
+
```
the test phase cannot have worked... as the environment is not setup correctly when things get compiled. You should instead add the following before ```test.run```:
```
    pre-test {
        test.env    PYTHONPATH=[glob -nocomplain ${worksrcpath}/build/lib*]
    }
```

additionally, please: use  ```${python.bin}``` for the ```test.cmd```

essandess replied:
Thanks done. 🦃🦃🦃

essandess commented:
@cjones051073 Do you know why py38-pytorch fails to build?

I have verified that the Python version 3.7 build of the parent py-allennlp works both locally and within the build bot.

Have I specified something incorrect in the parent Portfile for py-allennlp? Is this is known issue?

See: https://paste.z0k.xyz/aa3770fbe290
```
[ 96%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/Device.cpp.o
[1m/opt/local/var/macports/build/_Users_runner_runners_2.160.1_work_1_s_python_py-pytorch/py38-pytorch/work/pytorch-1.3.1/test/cpp/api/serialize.cpp:333:25: [0m[0;1;35mwarning: [0m[1mcomparison of integers of different signs: 'int64_t' (aka 'long long') and 'std::__1::vector<at::Tensor, std::__1::allocator<at::Tensor> >::size_type' (aka 'unsigned long') [-Wsign-compare][0m
  for (int64_t i = 0; i < x_vec.size(); i++) {
[0;1;32m                      ~ ^ ~~~~~~~~~~~~
[0m[1m/opt/local/var/macports/build/_Users_runner_runners_2.160.1_work_1_s_python_py-pytorch/py38-pytorch/work/pytorch-1.3.1/torch/csrc/Device.cpp:185:3: [0m[0;1;31merror: [0m[1mcannot initialize a member subobject of type 'Py_ssize_t' (aka 'long') with an rvalue of type 'nullptr_t'[0m
  nullptr,                                     /* tp_print */
[0;1;32m  ^~~~~~~
[0m[ 96%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/Dtype.cpp.o
[1m/opt/local/var/macports/build/_Users_runner_runners_2.160.1_work_1_s_python_py-pytorch/py38-pytorch/work/pytorch-1.3.1/torch/csrc/Dtype.cpp:65:3: [0m[0;1;31merror: [0m[1mcannot initialize a member subobject of type 'Py_ssize_t' (aka 'long') with an rvalue of type 'nullptr_t'[0m
  nullptr,                                     /* tp_print */
[0;1;32m  ^~~~~~~
[0m1 error generated.
make[2]: *** [caffe2/torch/CMakeFiles/torch_python.dir/csrc/Device.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
[ 96%] Building CXX object test_api/CMakeFiles/test_api.dir/static.cpp.o
1 error generated.
```

cjones051073 commented:
I was a little premature in adding py38 support to pytorch. I thought it was there, but turns out not quite yet. Just ignore for now. I will likely remove py38 from the Portfile in the meantime.

essandess commented:
> I was a little premature in adding py38 support to pytorch. I thought it was there, but turns out not quite yet. Just ignore for now. I will likely remove py38 from the Portfile in the meantime.

@cjones051073 Okay thank you. That knowledge saves some downstream debugging.

Now only there were any way at all to run GPUs on macOS. See https://github.com/RadeonOpenCompute/ROCm/issues/262 for part of the sad story.

kencu commented:
I'm looking to kill off ancient PRs that aren't going anywhere -- is this one?

essandess commented:
It’s waiting on upstream to release RC 1.0.0: https://github.com/allenai/allennlp/releases.

kencu commented:
Something with that name has been released 18 hours ago, i note, FYI. OK, leave this open then.

kencu commented:
I note v1.1.0 has been released now, if that is what this PR has been waiting for all this time.

essandess commented:
This PR is now dependent on https://github.com/macports/macports-ports/pull/8783

essandess commented:
@kencu To the best of my knowledge, this PR is ready to merge.

> I note v1.1.0 has been released now, if that is what this PR has been waiting for all this time.

kencu commented:
@reneeotten -- you had concerns about this (a year ago!). Any concerns about committing this now?

kencu commented:
I will commit this tomorrow unless there are any objections raised

essandess commented:
> These additions would be another nice addition to the MacPorts portfolio.
> 
> Apart from the one conflict - py-flaky - is this PR close to merge-ready?

I've updated all submissions to their latest versions and resolved the conflict. To the best of my knowledge it's ready to merge.

reneeotten commented:
@essandess I tried this PR locally, but it is still missing some dependencies and a few other pieces. For example, the port `jsonnet` does not build for me and if you follow the link that @mf2k gave above about this you should see this locally yourself as well. Another thing there is that you ideally should use the `nlohman-json` port as it is present in MacPorts already instead of using the bundled one. 

To try and get this merged I will commit the Python ports form this PR that are ready-to-go and then you can rebase and work on the others. Ideally the CI systems would finish as well with a smaller subset of ports.

reneeotten commented on python/py-catalogue/Portfile:
```diff
@@ -6,11 +6,12 @@ PortGroup           python 1.0
 epoch               1
 # Don't upgrade past py-spaCy's highest supported version.
-github.setup        explosion catalogue 1.0.0 v
+github.setup        explosion catalogue 2.0.1 v
```
this version is above the one supported by the py-spaCy's port in MacPorts

essandess replied:
Thanks done.

reneeotten commented on python/py-catalogue/Portfile:
```diff
@@ -20,18 +21,20 @@ license             MIT
+                    because the name is a simple string. If you \
+                    instead saved the function, you'd have to use \
+                    Pickle for serialization, which has many \
+                    drawbacks.
```
also, please don't make unnecessary white-space changes (or do so in a separate commit)

essandess replied:
Thanks done.

reneeotten commented on devel/jsonnet/Portfile:
```diff
@@ -0,0 +1,37 @@
+    reinplace -E \
+        {s/^([[:space:]]+(cp|mkdir)[[:space:]]+.+[[:space:]]+)\$\(PREFIX\)/\1$(DESTDIR)$(PREFIX)/g} \
+        ${worksrcpath}/Makefile
+}
```
Warning: reinplace s/^([[:space:]]+(cp|mkdir)[[:space:]]+.+[[:space:]]+)\$\(PREFIX\)/\1$(DESTDIR)$(PREFIX)/g didn't change anything

reneeotten replied:
please use a patch file to make required changes instead of cryptic reinplace statements

essandess replied:
Thanks for the working patch file below. Done.

reneeotten commented on devel/jsonnet/Portfile:
```diff
@@ -0,0 +1,37 @@
+                    sha256  9f4450ff784952a0f5c593ce7dabd83cbda2eed62a4b01cc2ba507b32483c8dd \
+                    size    21859800
+
+use_configure       no
```
this port is not using TheRightCompiler. Ideally you would use the `makefile` PG or perhaps switch to CMake. In anycase, the port in its current form does not build for me locally.

essandess replied:
@reneeotten Thanks for all these corrections.

I'm trying to get `jsonnet` to compile with cmake. It hits a link error with `ld: symbol(s) not found`.

It looks like there's an extra directory `${cmake.build_dir}/lib` with some compiled libraries.

I can't tell if this is a bug in the cmake build, or I just need to somehow add a `-L${cmake.build_dir}/lib` to the link command.

How do I do this in MacPorts?  If I do this, the cmake configure phase fails because the directory doesn't yet exist:
```
configure.ldflags-append \
                    ${cmake.build_dir}/lib

```

Failing this, it looks like an upstream issue.

essandess replied:
P.S. See 
[main.log](https://github.com/macports/macports-ports/files/5982275/main.log)

essandess replied:
Got this working per makefile as you describe below. Upstream's cmake appears pretty broken in several places.

Done.

reneeotten commented on devel/jsonnet/Portfile:
```diff
@@ -0,0 +1,37 @@
+
+long_description    A data templating language for app and tool developers
+
+homepage            http://jsonnet.org/
```
please use https

reneeotten commented on python/py-jsonnet/Portfile:
```diff
@@ -0,0 +1,43 @@
+
+long_description    A data templating language for app and tool developers
+
+homepage            http://jsonnet.org/
```
please use https

essandess replied:
Thanks done.

reneeotten commented on python/py-tensorboardX/Portfile:
```diff
@@ -0,0 +1,55 @@
+                    port:py${python.version}-six
+
+    depends_test-append \
+                    port:py${python.version}-crc32c \
```
this dependency isn't available in MacPorts

essandess replied:
Thanks done.

reneeotten commented:
@essandess I have committed the ports from this PR that were working correctly (made a few minor changes here and there). There are some issues with the remaining ones here, but it's a small number now so if you can take care of my review comments and the ```lint``` warnings that the CI will give, I am happy to take another look and get this finalized.

reneeotten commented on python/py-allennlp/Portfile:
```diff
@@ -0,0 +1,81 @@
+    depends_test-append \
+                    port:py${python.version}-flaky \
+                    port:py${python.version}-pytest \
+                    port:py${python.version}-responses
```
the `py39-allennlp` has many missing dependencies: 
```
py39-allennlp
  Error: Unknown dependency: py39-boto3
  Error: Unknown dependency: py39-nltk
  Error: Unknown dependency: py39-parsimonious
  Error: Unknown dependency: py39-transformers
  Error: Unknown dependency: py39-unidecode
  Warning: Dependency port:py39-scipy specified multiple times in depends_run
```

essandess replied:
Thanks done.

reneeotten commented:
@essandess your force-push seems not to have changed anything compared to what I pushed already yesterday... or?

essandess commented:
> @essandess your force-push seems not to have changed anything compared to what I pushed already yesterday... or?

I hope not! I rebased and deleted the commits you added. Did I inadvertently delete stuff too?

reneeotten commented:
> > @essandess your force-push seems not to have changed anything compared to what I pushed already yesterday... or?
> 
> I hope not! I rebased and deleted the commits you added. Did I inadvertently delete stuff too?

I am not sure what your intention was - I pushed yesterday the remaining ports to this PR and left comments about things that still needed to be changed. Then I noticed that you force-pushed again, without taking care of any of the comments/requests. So I guess I was just confused about the purpose of your force-push.

essandess commented:
My intent is a to catch up to your edits.

I am stuck at being able to build `jsonnet` with cmake. Please see my response above.

reneeotten commented:
> My intent is a to catch up to your edits.
> 
> I am stuck at being able to build `jsonnet` with cmake. Please see my response above.

okay, I think it can be done by patching the Makefile and using the makefile PG; see [here](https://github.com/reneeotten/macports-ports/commit/3febf003954d221181bf2e7d2d0ea37c5cd6cdda). I think this adds all the flags set by MacPorts, uses the `nlohmann-json` port and seems to work for me. 

The question is how to add the Python bindings; if you add them in a separate `py-jsonnet` file you'll have to duplicate the `jsonnet` Portfile since you first need to compile the thing again and then build the Python bindings. Just doing a `make` in the `pre-build` as you've done in this PR doesn't work correctly since that doesn't use TheRightCompiler and nor sets the correct flags.

I *think* it might be easier in this case to add the pyXY subports to `devel/jsonnet` so that you don't have to replicate the build of `jsonnet` but only can do the `python setup.py .....` in the subports. I tried a little tonight, but can't figure out yet how to do it... and to be honest I don't really much more time to invest in it.

essandess commented:
Thanks, I have `jsonnet` build now and the tests and run successfully, but the libraries are all `.so`. Is this okay?
```bash
port contents jsonnet
Port jsonnet contains:
  /opt/local/bin/jsonnet
  /opt/local/bin/jsonnetfmt
  /opt/local/include/libjsonnet++.h
  /opt/local/include/libjsonnet.h
  /opt/local/include/libjsonnet_fmt.h
  /opt/local/lib/libjsonnet++.so
  /opt/local/lib/libjsonnet++.so.0
  /opt/local/lib/libjsonnet++.so.0.17.0
  /opt/local/lib/libjsonnet.so
  /opt/local/lib/libjsonnet.so.0
  /opt/local/lib/libjsonnet.so.0.17.0
```

> okay, I think it can be done by patching the Makefile and using the makefile PG; see [here](https://github.com/reneeotten/macports-ports/commit/3febf003954d221181bf2e7d2d0ea37c5cd6cdda). I think this adds all the flags set by MacPorts, uses the `nlohmann-json` port and seems to work for me.

essandess commented:
> Thanks, I have `jsonnet` build now and the tests all run successfully, but the libraries are all `.so`. Is this okay?
> 
> ```shell
> port contents jsonnet
> Port jsonnet contains:
>   /opt/local/bin/jsonnet
>   /opt/local/bin/jsonnetfmt
>   /opt/local/include/libjsonnet++.h
>   /opt/local/include/libjsonnet.h
>   /opt/local/include/libjsonnet_fmt.h
>   /opt/local/lib/libjsonnet++.so
>   /opt/local/lib/libjsonnet++.so.0
>   /opt/local/lib/libjsonnet++.so.0.17.0
>   /opt/local/lib/libjsonnet.so
>   /opt/local/lib/libjsonnet.so.0
>   /opt/local/lib/libjsonnet.so.0.17.0
> ```
> 
> > okay, I think it can be done by patching the Makefile and using the makefile PG; see [here](https://github.com/reneeotten/macports-ports/commit/3febf003954d221181bf2e7d2d0ea37c5cd6cdda). I think this adds all the flags set by MacPorts, uses the `nlohmann-json` port and seems to work for me.

mf2k commented on devel/googletest/Portfile:
```diff
@@ -0,0 +1,26 @@
+                    Testing Technology team with Google's specific \
+                    requirements and constraints in mind.
+
+github.master_sites ${github.homepage}/archive/
+distfiles           release-${version}${extract.suffix}
```
`distfiles` should never be set directly. Both of these lines will be correctly set by the GitHub portgroup. Just tell it to grab from archives.

`github.tarball_from archives
`

essandess replied:
None of the `github.tarball_from` options work in this case because the tarball is not named using `${git.branch}`. Please advise:
```
        switch ${args} {
            downloads {
                github.master_sites https://github.com/downloads/${github.author}/${github.project}
            }
            releases {
                github.master_sites ${github.homepage}/releases/download/${git.branch}
            }
            archive {
                github.master_sites ${github.homepage}/archive/${git.branch}
            }
            tarball {
		github.master_sites ${github.homepage}/tarball/${git.branch}
            }
            tags {
                return -code error "the value \"tags\" is deprecated for github\
.tarball_from. Please use \"tarball\" instead."
            }
```

essandess replied:
nm—I see I had the tag prefix set correctly. Thanks, done.

mf2k commented on python/py-allennlp/Portfile:
```diff
@@ -0,0 +1,81 @@
+                    port:py${python.version}-pytorch \
+                    port:py${python.version}-requests \
+                    port:py${python.version}-responses \
+                    port:py${python.version}-scipy \
+                    port:py${python.version}-scikit-learn \
+                    port:py${python.version}-scipy \
```
per lint, `port:py${python.version}-scipy` appears twice here.

essandess replied:
Fixed, thanks, along with a bunch of other dependencies that have changes since this PR was posted originally.

essandess commented:
I believe that a couple of these submissions need a `depends_lib` for `setuptools`; I’ll correct these…

reneeotten commented on python/py-catalogue/Portfile:
```diff
@@ -59,5 +60,10 @@ if {${name} ne ${subport}} {
+     test.cmd        py.test-${python.branch}
+     test.target
+     test.env        PYTHONPATH=${worksrcpath}/build/lib
+
```
why duplicating this, it's already set above the `post-destroot` part it seems

essandess replied:
Oops, thanks, fixed.

reneeotten commented on devel/googletest/Portfile:
```diff
@@ -0,0 +1,25 @@
+
+checksums           rmd160  f2f7f96adf1ab8aae3f6204fdba3ff150fd4396b \
+                    sha256  9dc9157a9a1551ec7a7e43daea9a694a0bb5fb8bec81235d8a1e6ef64c716dcb \
+                    size    904349
```
please drop this commit, the port is already present as `devel/gtest`

essandess replied:
Oops, thanks, done.

reneeotten commented on python/py-pytest-benchmark/Portfile:
```diff
@@ -0,0 +1,42 @@
+    depends_run-append \
+                    port:py${python.version}-cpuinfo \
+                    port:py${python.version}-pytest \
+                    port:py${python.version}-tox
```
I would be surprised to have `tox` as a runtime dependency for a pytest plugin... are you sure?

essandess replied:
Thanks—cut-and-paste error fixed.

essandess commented:
This built successfully on both 10.15 and 11.0 previous to the last push with minor mods. AFAIK it’s ready to go.

essandess commented:
Whoops forgot to commit an addition of `py37-lmdb`. This should build now…

> This built successfully on both 10.15 and 11.0 previous to the last push with minor mods. AFAIK it’s ready to go.

essandess commented:
@reneeotten @mf2k I believe that this is ready to merge. The CI issues are from the large number of builds, not the Portfile code.

mf2k commented:
@reneeotten : ping?

mf2k commented:
@essandess : It looks like there is a conflict to fix.

essandess commented:
> @essandess : It looks like there is a conflict to fix.

Fixed thx

essandess commented:
> I would say yes it would.

Thanks done.

reneeotten commented:
> > I would say yes it would.
> 
> Thanks done.

well.... now there are ~40 different PRs with one commit that do not build either because the dependencies are now in a separate commit. Yes, having smaller PRs is great - but they should still build, otherwise I'd say it's very unlikely that someone will merge them...

essandess commented:
> > > I would say yes it would.
> > 
> > 
> > Thanks done.
> 
> well.... now there are ~40 different PRs with one commit that do not build either because the dependencies are now in a separate commit. Yes, having smaller PRs is great - but they should still build, otherwise I'd say it's very unlikely that someone will merge them...

What's the alternative?

cjones051073 commented:
We just need to start at the beginning, the lowest dependency, review and merge that and then move up the tree as we can testing them.

essandess commented:
I believe that this is ready to merge. The only CI issues are timeouts, unsurprising for a port with both tensorflow and pytorch as depencies.

essandess commented:
> I believe that this is ready to merge. The only CI issues are timeouts, unsurprising for a port with both tensorflow and pytorch as depencies.

@reneeotten I believe that this is ready to merge.

reneeotten commented:
> > I believe that this is ready to merge. The only CI issues are timeouts, unsurprising for a port with both tensorflow and pytorch as depencies.
> 
> @reneeotten I believe that this is ready to merge.

It seems all other ports are present indeed now and this Portfile looks fine after a quick glance... Can you please update the port to the latest upstream release and I'll merge after that.

essandess commented:
> It seems all other ports are present indeed now and this Portfile looks fine after a quick glance... Can you please update the port to the latest upstream release and I'll merge after that.

Thank you, done.

essandess commented:
> > It seems all other ports are present indeed now and this Portfile looks fine after a quick glance... Can you please update the port to the latest upstream release and I'll merge after that.
> 
> Thank you, done.

The CI looks busted at the moment but I confirm it installs and runs on macOS 11 and previously on macOS 10.15.

reneeotten commented on python/py-allennlp/Portfile:
```diff
@@ -0,0 +1,77 @@
+                    port:py${python.version}-responses
+
+    test.run        yes
+    test.env        PYTHONPATH=${worksrcpath}/build/lib
```
sorry, one more question... The Portfile doesn't set `supported_archs noarch`, so supposedly some code gets compiled here - correct? If so, shouldn't you instead set:
```
    pre-test {
        test.env    PYTHONPATH=[glob -nocomplain ${worksrcpath}/build/lib*]
    }
``` 

did you confirm that the test-suite runs by running the tests without having the port already installed?

essandess replied:
Yes, I did. The port install some `cpython` stuff, but I don't know if the `supported_archs` should be set.

essandess replied:
@reneeotten I added `supported_archs noarch`. I believe that this is ready to merge.

essandess commented:
@reneeotten @mf2k @cjones051073 

May we please merge this PR? It's been active longer than the pandemic. It's been through not one but two major version updates and numerous minor ones.

Maintaining this PR takes more work than a merged Portfile.

Let's declare victory and address any future edits in the next version update.
